internal/discovery/namespace_filter.go

```diff
@@ -0,0 +1,126 @@
+package discovery
+
+import (
+	"errors"
+
+	"github.com/newrelic/nri-kubernetes/v3/internal/config"
+	"github.com/newrelic/nri-kubernetes/v3/internal/storer"
+
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	listersv1 "k8s.io/client-go/listers/core/v1"
+)
+
+// NamespaceFilterer provides an interface to filter namespaces.
+type NamespaceFilterer interface {
+	IsAllowed(namespace string) bool
+}
+
+// NamespaceFilter is a struct holding pointers to the config and the namespace lister.
+type NamespaceFilter struct {
+	c      *config.Config
+	lister listersv1.NamespaceLister
+	stopCh chan<- struct{}
+	storer storer.Storer
+}
+
+// NewNamespaceFilter inits the namespace lister and returns a new NamespaceFilter.
+func NewNamespaceFilter(c *config.Config, client kubernetes.Interface, storer storer.Storer, options ...informers.SharedInformerOption) *NamespaceFilter {
+	stopCh := make(chan struct{})
+
+	factory := informers.NewSharedInformerFactoryWithOptions(client, defaultResyncDuration, options...)
+
+	lister := factory.Core().V1().Namespaces().Lister()
+
+	factory.Start(stopCh)
+	factory.WaitForCacheSync(stopCh)
+
+	return &NamespaceFilter{
+		c:      c,
+		lister: lister,
+		stopCh: stopCh,
+		storer: storer,
+	}
+}
+
+// IsAllowed checks given any namespace, if it's allowed to be scraped by using the NamespaceLister
+func (nf *NamespaceFilter) IsAllowed(namespace string) bool {
+	// By default, we scrape every namespace.
+	if nf.c.NamespaceSelector == nil {
+		return true
+	}
+
+	// Check if the namespace is already in the cache.
+	var allowed bool
+	if _, err := nf.storer.Get(namespace, &allowed); err == nil {
+		return allowed
+	}
```

Comment:
Using the same `storer` to cache, a different approach which would be availing ourselves of the `NamespaceFilter` interface definition. I'm thinking of something like this:

```go
type CachedNamespaceFilter struct {
    nsFilter NamespaceFilter
    cache storer.Storer
}


func (nf *CachedNamespaceFilter) IsAvailable(namespace string) bool {
// checks cache and calls the underlying NamespaceFilter if the result is not found
}
```

It would be possible to keep tests as they where and test the whole cache thing separately.

WDYT?

Reply:
Updated in 7158e361f86e9713a3d2497ab1ff194a9d33b044.